OpenRA.Game/Server/PlayerMessageTracker.cs

```diff
@@ -0,0 +1,72 @@
+#region Copyright & License Information
+/*
+ * Copyright 2007-2022 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+
+namespace OpenRA.Server
+{
+	class PlayerMessageTracker
+	{
+		readonly Dictionary<int, List<long>> messageTracker = new Dictionary<int, List<long>>();
+		readonly Server server;
+		readonly Action<Connection, int, int, byte[]> disableChatEntry;
+		readonly Action<Connection, string, Dictionary<string, object>> sendLocalizedMessageTo;
+
+		public PlayerMessageTracker(Server server, Action<Connection, int, int, byte[]> disableChatEntry, Action<Connection, string, Dictionary<string, object>> sendLocalizedMessageTo)
+		{
+			this.server = server;
+			this.disableChatEntry = disableChatEntry;
+			this.sendLocalizedMessageTo = sendLocalizedMessageTo;
+		}
+
+		public void DisableChat(Connection conn, uint time)
+		{
+			disableChatEntry(conn, 0, 0, new Order("DisableChatEntry", null, false) { ExtraData = time }.Serialize());
+		}
+
+		public bool IsWritable(Connection conn, string chatDisabled)
+		{
+			if (!messageTracker.ContainsKey(conn.PlayerIndex))
+				messageTracker.Add(conn.PlayerIndex, new List<long>());
+
+			var isAdmin = server.GetClient(conn)?.IsAdmin ?? false;
+			var time = conn.ConnectionTimer.ElapsedMilliseconds;
+			var tracker = messageTracker[conn.PlayerIndex];
+			tracker.RemoveAll(t => t + server.Settings.FloodLimitInterval < time);
+
+			if (!isAdmin && time < server.Settings.FloodLimitJoinCooldown)
+			{
+				var remaining = (server.Settings.FloodLimitJoinCooldown - time + 999) / 1000;
```

Comment:
for extra bonus points you can extract this calculation to a method and use it here and on line 55.

Reply:
how should the method be called though? `TryDisableChatOnStart`?

Reply:
I think you misunderstood i meant `var remaining = (server.Settings.FloodLimitJoinCooldown - time + 999) / 1000;, `.
So it should be like this.

```cs
			long CalculateRemaining(long cooldown)
			{
				return (cooldown - time + 999) / 1000;
			}

			if (!isAdmin && time < server.Settings.FloodLimitJoinCooldown)
			{
				var remaining = CalculateRemaining(server.Settings.FloodLimitJoinCooldown);
				sendLocalizedMessageTo(conn, chatDisabled, Translation.Arguments("remaining", remaining));
				return false;
			}

			if (tracker.Count >= server.Settings.FloodLimitMessageCount)
			{
				var remaining = CalculateRemaining(tracker[0] + server.Settings.FloodLimitCooldown);
				sendLocalizedMessageTo(conn, chatDisabled, Translation.Arguments("remaining", remaining));
				return false;
			}
```

Reply:
ah I completely missed the point.

I don't think it's worth extracting as I need both `cooldown` and `time` for the function. At that point I'm barely extracting anything

Reply:
implemented as local function